pkg/cli/bootstrap.go

```diff
@@ -0,0 +1,262 @@
+// Copyright 2021 Outreach Corporation. All Rights Reserved.
+
+// Description: This file contains cli functions used in bootstrap
+// and eventually in stencil.
+package cli
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"os/user"
+	"runtime"
+	"runtime/debug"
+	"strings"
+	"syscall"
+
+	"github.com/getoutreach/gobox/pkg/app"
+	"github.com/getoutreach/gobox/pkg/cfg"
+	"github.com/getoutreach/gobox/pkg/exec"
+	"github.com/getoutreach/gobox/pkg/log"
+	"github.com/getoutreach/gobox/pkg/secrets"
+	"github.com/getoutreach/gobox/pkg/trace"
+	"github.com/getoutreach/gobox/pkg/updater"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli/v2"
+	"gopkg.in/yaml.v2"
+)
+
+// UpdateExitCode is the exit code returned when an update ocurred
+const UpdateExitCode = 5
+
+// overrideConfigLoaders fakes certain parts of the config that usually get pulled
+// in via mechanisms that don't make sense to use in CLIs.
+func overrideConfigLoaders(honeycombAPIKey, dataset string, tracingDebug bool) {
+	var fallbackSecretLookup func(context.Context, string) ([]byte, error)
+	fallbackSecretLookup = secrets.SetDevLookup(func(ctx context.Context, key string) ([]byte, error) {
+		if key == "APIKey" {
+			return []byte(honeycombAPIKey), nil
+		}
+
+		return fallbackSecretLookup(ctx, key)
+	})
+
+	fallbackConfigReader := cfg.DefaultReader()
+	cfg.SetDefaultReader(func(fileName string) ([]byte, error) {
+		if fileName == "trace.yaml" {
+			traceConfig := &trace.Config{
+				Honeycomb: trace.Honeycomb{
+					Enabled: true,
+					APIHost: "https://api.honeycomb.io",
+					APIKey: cfg.Secret{
+						Path: "APIKey",
+					},
+					Debug:         tracingDebug,
+					Dataset:       dataset,
+					SamplePercent: 100,
+				},
+			}
+			b, err := yaml.Marshal(&traceConfig)
+			if err != nil {
+				panic(err)
+			}
+			return b, nil
+		}
+
+		return fallbackConfigReader(fileName)
+	})
+}
+
+// intPtr turns an int into a *int
+func intPtr(i int) *int {
+	return &i
+}
+
+// funcPtr turns a func into a *func
+func funcPtr(fn func()) *func() {
+	return &fn
+}
+
+// urfaveRegisterShutdownHandler registers a signal notifier that translates various term
+// signals into context cancel
+func urfaveRegisterShutdownHandler(cancel context.CancelFunc) {
+	// handle ^C gracefully
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM, syscall.SIGHUP)
+	go func() {
+		<-c
+		signal.Reset()
+		cancel()
+	}()
+}
+
+// setupTracer sets up a root trace for the CLI and initializes the tracer
+func setupTracer(ctx context.Context, name string) context.Context {
+	if err := trace.InitTracer(ctx, name); err != nil {
+		fmt.Println(err)
+		return ctx
+	}
+	return trace.StartTrace(ctx, name)
+}
+
+// setupPanicHandler sets up a panic handler for CLIs
+func setupPanicHandler(exitCode *int) {
```

Comment:
lol.  I think we can make this a bit more readable by creating a temporary struct to hold all the state and moving all these top-level functions as methods on that struct..

i.e:

```golang

type cli struct {
    exitCode int
    cleanup func()
}

func (c *cli) setupPanicHandler() {
       if r := recover(); r != nil {
          c.exitCode = 2
       }
}
```